aspects/transaction_test.go

```diff
@@ -0,0 +1,328 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspects_test
+
+import (
+	"errors"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/aspects"
+)
+
+type transactionTestSuite struct{}
+
+var _ = Suite(&transactionTestSuite{})
+
+type witnessReadWriter struct {
+	readCalled     int
+	writeCalled    int
+	bag            aspects.JSONDataBag
+	writtenDatabag aspects.JSONDataBag
+}
+
+func (w *witnessReadWriter) read() (aspects.JSONDataBag, error) {
+	w.readCalled++
+	return w.bag, nil
+}
+
+func (w *witnessReadWriter) write(bag aspects.JSONDataBag) error {
+	w.writeCalled++
+	w.writtenDatabag = bag
+	return nil
+}
+
+func (s *transactionTestSuite) TestSet(c *C) {
+	bag := aspects.NewJSONDataBag()
+	witness := &witnessReadWriter{bag: bag}
+	schema := aspects.NewJSONSchema()
+	tx, err := aspects.NewTransaction(witness.read, witness.write, schema)
+	c.Assert(err, IsNil)
+	c.Assert(witness.readCalled, Equals, 1)
+
+	err = tx.Set("foo", "bar")
+	c.Assert(err, IsNil)
+	c.Assert(witness.writeCalled, Equals, 0)
+
+	var value interface{}
+	err = witness.writtenDatabag.Get("foo", &value)
+	c.Assert(err, FitsTypeOf, &aspects.FieldNotFoundError{})
```

Comment:
This seems a little removed from the test objective. Perhaps `c.Assert(witness.writtenDatabag, IsNil)` will suffice?

Reply:
It tests that the field isn't set by checking the error is a FieldNotFound. Maybe I'm missing the point, but it seems appropriate

Reply:
It's a nitpick, just meant that `c.Assert(witness.writeCalled, Equals, 0)` already ensures that `w.writtenDatabag` does not get updated, so lines 63-65 seems superfluous.